c_cpp_guide.md

```diff
@@ -0,0 +1,115 @@
+# Compiling C/C++ to Ewasm
+
+First an introduction, then a basic step-by-step guide, then advanced things. Warning: the Ewasm spec and tools below are subject to change.
+
+## Introduction
+
+An Ewasm contract is a WebAssembly module with the following restrictions:
+
+- The module's imports must be among the [Ewasm helper functions](https://github.com/ewasm/design/blob/master/eth_interface.md) which resemble EVM opcodes to interact with the client.
+- The module's exports must be a `main` function which takes no arguments and returns nothing, and the `memory` of the module.
+- The module may not use floats or other [sources of non-determinism](https://github.com/WebAssembly/design/blob/master/Nondeterminism.md).
+
+When writing Ewasm contracts in C/C++, one should bear in mind the following caveats:
+
+1. WebAssembly is still primitive and [lacks features](https://github.com/WebAssembly/design/blob/master/FutureFeatures.md). For example, WebAssembly lacks support for exceptions and we have no way to do system calls in Ewasm. Compilers and libraries are still primitive. For example, we have a patched version of libc to allow `malloc`, but the patches are not yet enough for `std::vector` because other memory managment calls are unavailable. But perhaps any memory management beyond memory allocation may be unwanted for Ewasm contracts since it costs gas. This situation will improve as WebAssembly, compilers, and libraries mature.
```

Comment:
> But perhaps any memory management beyond memory allocation may be unwanted for Ewasm contracts since it costs gas.

I don't think this statement belongs here, it leaves the reader with too many unanswered questions. I think the best thing to do would be to compile a list of open design questions in one place (not specific to C/C++) and provide a link to it somewhere in this doc. A link to open issues on the ewasm/design repo might suffice for now if we don't have a more mature doc.

To make this more helpful and constructive, it would be nice to conclude this section by saying something along the lines of, "For now, to work around these issues, ensure that you only use basic structs and `malloc` calls" (or whatever the advice should be). "To join the conversation and contribute to ongoing Ewasm interface design, see X" (link to another doc).

> This situation will improve as WebAssembly, compilers, and libraries mature.

Consider dropping, I don't think this is critical.

Reply:
Agreed that a guide is not a place for design discussions. I overhauled this paragraph.